Add Zonal Model Builder

AnanEzeiry opened the pull request:
## Todo
- [x] Rebase
- [x] Fix ruff / formatting
- [x] Add tests
- [x] Add documentation

Commit:
[WIP] Add Zonal Rate Model Builder

lieres commented on CADETProcess/modelBuilder/ZRMBuilder.py:
```diff
@@ -0,0 +1,135 @@
+binding_model.desorption_rate = [1]
+binding_model.capacity = 1.2e3
+
+#In both scenarios axial or radial configuration, we defined one inlet that goes into one cstr
```
defined -> define

lieres replied:
cstr -> CSTR

AnanEzeiry replied:
Done

lieres commented on CADETProcess/modelBuilder/ZRMBuilder.py:
```diff
@@ -0,0 +1,135 @@
+#In both scenarios axial or radial configuration, we defined one inlet that goes into one cstr
+inlet = Inlet(component_system, name='inlet')
+volumetric_flow_rate = 1e-6
+inlet.flow_rate = volumetric_flow_rate #flow rate is constant assuming constant liquid density 
```
flow rate could vary due to other reasons

AnanEzeiry replied:
Added

lieres commented on CADETProcess/modelBuilder/ZRMBuilder.py:
```diff
@@ -0,0 +1,135 @@
+            self.flow_sheet.add_unit(zone)
+            self.zones.append(zone)
+
+        # Create symmetric CSTRs' volumes
```
for later purpose: define CSTRs before and behind the membrane stack separately

AnanEzeiry replied:
Defined

lieres commented on CADETProcess/modelBuilder/ZRMBuilder.py:
```diff
@@ -0,0 +1,135 @@
+                    f'zone{i+1}': zone_fraction
+                }
+            )
+        # Inlet and outlet connections
```
This is the only place where axial flow and radial flow differs. You can already include both variants with conditional statements.

AnanEzeiry replied:
It's not the previous block? I though this block is similar in both cases!

AnanEzeiry commented on CADETProcess/modelBuilder/ZRMBuilder.py:
```diff
@@ -0,0 +1,135 @@
+
+        # Set output states for flow splitting
+        total_area = sum(self.segments_area)
+        epsilon = 1e-12  # Prevent divide-by-zero
```
Epsilon, replaced by lowest_bound to avoid dividing by zero

lieres commented on CADETProcess/modelBuilder/AxialZRMBuilder.py:
```diff
@@ -0,0 +1,171 @@
+        Builder class for Axial Zoneal Rate Model (ZRM) with symmetric cstrs layout.
+
+        Args:
+            component_system: number of components.
```
The component system is more than just the number of components.

AnanEzeiry replied:
you mean that it's also include the type, e.g salt ?

lieres replied:
Yes

lieres commented on CADETProcess/modelBuilder/AxialZRMBuilder.py:
```diff
@@ -0,0 +1,171 @@
+    def Segments_ratio(self):
+        """Returns the surface area ratio of each zone."""
+        total_area = sum(self.segments_area)
+        if total_area == 0:
```
Line 71: Total area are cannot be 0 when all segment areas are >0.

AnanEzeiry replied:
Fixed

lieres commented on CADETProcess/modelBuilder/AxialZRMBuilder.py:
```diff
@@ -0,0 +1,171 @@
+
+    def build(self, length, porosity, axial_dispersion, volumes):
+        """Constructs the flow sheet."""
+        if len(volumes) not in (self.n, 2 * self.n):
```
Line 77: I would prefer two vectors for the CSTRs before and behind the membrane. If the second vector is empty, it is replaced by a copy of the first one. Otherwise, its length needs to be the same es of the first one.

AnanEzeiry replied:
If I understand you correctly, I did that into the same class within cstr_in and cstr_out, when the problem is symmetric then the cstr_out will be specified  automatically

lieres replied:
Yes, but I am referring to the interface.

AnanEzeiry replied:
I see

lieres commented on CADETProcess/modelBuilder/AxialZRMBuilder.py:
```diff
@@ -0,0 +1,171 @@
+
+        # Create membrane zones
+        for i in range(self.n):
+            zone = LumpedRateModelWithoutPores(self.component_system, f"zone{i+1}")
```
Line 94: If possible, allow to externally specify the transport model.

AnanEzeiry replied:
Done

AnanEzeiry commented on CADETProcess/modelBuilder/AxialZRMBuilder.py:
```diff
@@ -0,0 +1,171 @@
+from CADETProcess.processModel import ComponentSystem
+component_system = ComponentSystem(1) #number of components
+
+from CADETProcess.processModel import Langmuir
```
What do you mean by “through the ZRMBuilder interface” (line 3)? What I thought is that the binding_model is specified as a ZRMBuilder input, and its parameters are specified externally, independent of ZRMBuilde.

lieres replied:
That seems like a valid solution.

schmoelder commented:
Hi Anan, I added the description of the PR. Please proceed with the missing to-dos before we can merge this. Let us know if you need help.

AnanEzeiry commented on CADETProcess/modelBuilder/ZRMBuilder.py:
```diff
@@ -0,0 +1,135 @@
+        self.segments_area = segments_area
+        self.n = len(segments_area)
+        self.binding_model = binding_model
+        self.name = name or 'AxialZRMBuilder'
```
The current script addresses both axial and radial configurations

AnanEzeiry commented on CADETProcess/modelBuilder/ZRMBuilder.py:
```diff
@@ -0,0 +1,135 @@
+        self.flow_sheet.add_unit(self.inlet, feed_inlet=True)
+        self.flow_sheet.add_unit(self.outlet)
+
+        # Create and configure zones
```
Fixed

schmoelder committed:
Add parameter-parmeter dependency module

schmoelder committed:
Add parameter-parmeter dependency module

schmoelder committed:
Add radial flow models

schmoelder committed:
Add radial flow models

schmoelder committed:
Use class name for ParameterWrapper

schmoelder committed:
Use class name for ParameterWrapper

schmoelder commented on CADETProcess/modelBuilder/ZRMBuilder.py:
```diff
@@ -0,0 +1,290 @@
+        volumes_in: list[float],
+        volumes_out: list[float]
+    ) -> FlowSheet:
+        """Construct the flow sheet."""
```
Improve docstrings

schmoelder commented on CADETProcess/modelBuilder/ZRMBuilder.py:
```diff
@@ -0,0 +1,290 @@
+        if self.configuration not in ['axial', 'radial']:
+            raise ValueError("Configuration must be either 'axial' or 'radial'")
+
+    def build(
```
```suggestion
    def build_flow_sheet(
```

schmoelder commented on CADETProcess/modelBuilder/ZRMBuilder.py:
```diff
@@ -0,0 +1,290 @@
+        self._cycle_time = time
+
+    @property
+    def binding_model(self) -> None:
```
```suggestion
    def binding_model(self) -> BindingBaseClass:
```

schmoelder commented on CADETProcess/modelBuilder/ZRMBuilder.py:
```diff
@@ -0,0 +1,290 @@
+
+        with symmetric/Non-symmetric cstrs layout.
+
+        Args:
```
Be consistent with docstrings format. In CADET-Process, we use numpy style docstrings, not Google style.

AnanEzeiry replied:
Done

schmoelder commented on CADETProcess/modelBuilder/ZRMBuilder.py:
```diff
@@ -0,0 +1,290 @@
+        self.segments_area = segments_area
+        self.init_c = init_c
+        self.configuration = configuration.lower()
+        self._validate_inputs()
```
Generally, it's fine to have a validate_inputs method. However, I would recommend checking before assigning attributes. That probably means, your `_validate_inputs` method should take arguments and return True/False or raise an exception.

This also makes it easier to test the individual method.

schmoelder commented on CADETProcess/modelBuilder/ZRMBuilder.py:
```diff
@@ -0,0 +1,290 @@
+        """
+        Builder class for Axial/Radial Zonel Rate Model (ZRM).
+
+        with symmetric/Non-symmetric cstrs layout.
```
Fix formatting / Add more description. What are the different configurations / layouts.

AnanEzeiry replied:
Done

schmoelder commented on CADETProcess/modelBuilder/ZRMBuilder.py:
```diff
@@ -0,0 +1,290 @@
+    Attributes
+    ----------
+    segments_area : int
+        Surface areas of the segmented membrane.
```
Add missing attributes.

AnanEzeiry replied:
Done

schmoelder commented on CADETProcess/modelBuilder/ZRMBuilder.py:
```diff
@@ -0,0 +1,290 @@
+    @property
+    def binding_model(self) -> None:
+        """
+        Get the binding model associated with this builder.
```
Check docstring conventions for properties.

schmoelder commented on CADETProcess/modelBuilder/ZRMBuilder.py:
```diff
@@ -0,0 +1,290 @@
+        self.outlet = Outlet(component_system, name='outlet')
+
+    @property
+    def flow_sheet(self) -> FlowSheet:
```
do we need a flow_sheet property or should the ZRM only build a flow_sheet?

AnanEzeiry replied:
Yes, we need it since it is adjustable, for example, to add a PFR to account for the delay time

schmoelder commented on CADETProcess/modelBuilder/ZRMBuilder.py:
```diff
@@ -0,0 +1,290 @@
+        return self._flow_sheet
+
+    @property
+    def process(self)  -> Process:
```
do we need a process property or should the ZRM only build a process?

schmoelder commented on CADETProcess/modelBuilder/ZRMBuilder.py:
```diff
@@ -0,0 +1,290 @@
+
+    @cycle_time.setter
+    def cycle_time(self, time: float)  -> None:
+        self._cycle_time = time
```
Note, this setter does not automatically update the cycle time of the process.

schmoelder reviewed (commented):
We're slowly getting there. :nerd_face:

schmoelder commented on CADETProcess/modelBuilder/ZRMBuilder.py:
```diff
@@ -0,0 +1,290 @@
+
+    def _validate_inputs(self) -> None:
+        """Validate init inputs."""
+        if isinstance(self.flow_rate, str):
```
Note, type / value checking of the flow rate is automatically performed by the `Inlet` when you set its property.

schmoelder replied:
Just a note, if you were to test, don't only check for strings, but rather make sure it's a valid entry.

```suggestion
        if not np.isscalar(self.flow_rate):
```

schmoelder commented on CADETProcess/modelBuilder/ZRMBuilder.py:
```diff
@@ -0,0 +1,290 @@
+            membrane_segment.binding_model = binding_model
+
+    # returning the ratio of each zone membrane area from the total membrane area
+    def Segments_ratio(self) -> list[float]:
```
Upper case variables are usually reserved for class definitions.

```suggestion
    def segments_ratio(self) -> list[float]:
```

schmoelder commented on CADETProcess/modelBuilder/ZRMFlowSheetBuilder.py:
```diff
@@ -0,0 +1,290 @@
+        elif any(v_in < 0 for v_in in volumes_in):
+            raise ValueError("The cstr volume in volumes_in can not be negative float.")
+
+        # Proceed with the rest of the build logic here
```
Is this an open todo or can this be removed?

AnanEzeiry replied:
can you re-explain what do you mean here?

schmoelder commented on CADETProcess/modelBuilder/ZRMBuilder.py:
```diff
@@ -0,0 +1,290 @@
+        # Proceed with the rest of the build logic here
+
+        self._initialize_units(
+            rate_model,
```
Add type annotations

schmoelder commented on CADETProcess/modelBuilder/ZRMBuilder.py:
```diff
@@ -0,0 +1,290 @@
+            volumes_in,
+            volumes_out
+        )
+        self._connect_units()
```
Each method should do one thing only. Here the method implies that units are initialized, but instead you're connecting units (which is part of the flow sheet generation)

schmoelder replied:
Update: this is a duplicate method. See below. Remove!

AnanEzeiry committed:
Formatting and Docstrings

Consistent with the doctrisngs style, switching to numpy style

AnanEzeiry committed:
Lowercase the first letter of the method names

AnanEzeiry committed:
Importing NumPy as np to check that the flow rate is a scalar using np.isscalar

AnanEzeiry committed:
Renaming build method with build_flow_sheet

AnanEzeiry committed:
improved NumPy-style docstring for your build_flow_sheet method

AnanEzeiry committed:
Deleting tests comment

AnanEzeiry committed:
Ruff check

AnanEzeiry committed:
adding type annotations to initialize

schmoelder committed:
Refactor ZRMBuilder

schmoelder commented:
We refactored the entire module. It's much simpler now and builds a (partial) flow sheet. It now also works for arbitrary column models, binding models, reaction models etc. This way, it's easier to integrate other unit operations, such as PFRs into flow sheet without having to hard code anything. Note, inlets and outlets now also need to be manually added after calling `build_flow_sheet`.

Since this is effectively now just a "meta-unit", in the future, we might want to consider an improved behavior in CADET-Process where we don't any "FlowSheetBuilder" but can simply resolve this when translating to CADET-Core. The same is true for other systems such as the `CarouselBuilder` or the `CompartmentBuilder`.

Note, tests have not yet been updated. :sweat_smile:

AnanEzeiry committed:
Changing the name to ZRMFlowSheetBuilder

AnanEzeiry committed:
Refactoring the test to new ZRMFlowSheetBuilder and rename it to test_ZRMFlowSheetBuilder

AnanEzeiry committed:
changing the name to ZRMFlowSheetBuilder

AnanEzeiry committed:
this is the new test for ZRMFlowSheetBuilder, the last one was a mistake

AnanEzeiry committed:
old test deleted

AnanEzeiry committed:
Remving empty outlet cstr volume list

AnanEzeiry committed:
correct a typo

AnanEzeiry committed:
Remove unrelated modelBuilder files

AnanEzeiry commented:
After refactoring the ZRMBuilder into the ZRMFlowSheetBuilder module, a new adjusted test was created, uploaded, and executed successfully. New documentation is required and is currently in progress. 😉

AnanEzeiry committed:
Merge branch 'feature/radial_flow' into add_zonal_rate_model_builder

AnanEzeiry committed:
Adding membrane heights for radial configuration and membrane area for axial one

AnanEzeiry committed:
fix: add __all__ to ZRMFlowSheetBuilder to expose class via wildcard import

AnanEzeiry committed:
add a boolean check for radial transport model

AnanEzeiry committed:
Adding ZRMFlowSheetBuilder

AnanEzeiry committed:
adding __init__.py

AnanEzeiry committed:
fix issue: checking void_out if is not scalar or list, if scalar; spread it into list with number of zones items

AnanEzeiry committed:
update ZRMFlowSheetBuilder terminologies

In this update cstr wording has been replaced with void concept,
 together with the test

AnanEzeiry committed:
adding Module’s __all__ along with updating the autosummary

AnanEzeiry committed:
add executable ZRMFlowSheetBuilder documentation

In this commit a ZRMFlowSheetBuilder documentation is ready with
executable MyST cells, pushed together with figure, experimental data,
and citation
